crashed during startup

cube cannot start successfully if executing adjusting InitAppliedIndex during startup phase.

part of error log is as follows:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x104e9637c]

goroutine 882 [running]:
matrixone/pkg/vm/engine/aoe/dist.NewCubeDriverWithOptions.func8(0x14001d7ee70, 0x105469340, 0x14001987500, 0x0, 0x0, 0x0, 0x14000475c20)
        /Users/maomao/go/src/github.com/sukki37/matrixone/pkg/vm/engine/aoe/dist/cube_driver.go:205 +0x12c
github.com/matrixorigin/matrixcube/server.(*Application).AsyncExecWithGroupAndTimeout(0x140022423f0, 0x105469340, 0x14001987500, 0x1, 0x14001f71d28, 0x2540be400, 0x0, 0x0)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/server/server.go:168 +0x300
github.com/matrixorigin/matrixcube/server.(*Application).ExecWithGroup(0x140022423f0, 0x105469340, 0x14001987500, 0x1, 0x2540be400, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/server/server.go:125 +0xc8
matrixone/pkg/vm/engine/aoe/dist.(*driver).ExecWithGroup(...)
        /Users/maomao/go/src/github.com/sukki37/matrixone/pkg/vm/engine/aoe/dist/cube_driver.go:592
matrixone/pkg/vm/engine/aoe/dist.(*driver).GetSegmentedId(0x140022c73b0, 0x4, 0x105332c20, 0x14001f71d10, 0x0)
        /Users/maomao/go/src/github.com/sukki37/matrixone/pkg/vm/engine/aoe/dist/cube_driver.go:501 +0xa4
matrixone/pkg/vm/engine/aoe/dist.NewCubeDriverWithOptions.func7.1(0x4, 0x140011e2e00, 0x8, 0x8, 0x140011e2e08, 0x8, 0x8, 0x5, 0x1, 0x0, ...)
        /Users/maomao/go/src/github.com/sukki37/matrixone/pkg/vm/engine/aoe/dist/cube_driver.go:182 +0x50
github.com/matrixorigin/matrixcube/raftstore.getRaftConfig(0x41, 0xd, 0x140004af200, 0x1400014f080, 0x8)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/raftstore/peer_replica.go:437 +0x14c
github.com/matrixorigin/matrixcube/raftstore.newPeerReplica(0x140005c4b40, 0x140011b6d88, 0x41, 0x2d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1040b4101, ...)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/raftstore/peer_replica.go:189 +0x19c
github.com/matrixorigin/matrixcube/raftstore.createPeerReplica(0x140005c4b40, 0x140011b6d88, 0x105e31f58, 0x1053c0d20, 0x14002c1a180)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/raftstore/peer_replica.go:141 +0x88
github.com/matrixorigin/matrixcube/raftstore.(*store).startShards.func1(0x140011e2df0, 0xb, 0xb, 0x140005bc6c0, 0x34, 0x34, 0x1, 0x0, 0x0)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/raftstore/store.go:476 +0x3ac
github.com/matrixorigin/matrixcube/storage/pebble.(*Storage).Scan(0x1400166ad80, 0x105e045ee, 0x2, 0x2, 0x105e045f0, 0x2, 0x2, 0x14001a08f30, 0x140004bf300, 0x0, ...)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/storage/pebble/storage.go:171 +0x1b4
github.com/matrixorigin/matrixcube/raftstore.(*store).startShards(0x140005c4b40)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/raftstore/store.go:449 +0x184
github.com/matrixorigin/matrixcube/raftstore.(*store).Start(0x140005c4b40)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/raftstore/store.go:181 +0x1b0
github.com/matrixorigin/matrixcube/server.(*Application).Start(0x140022423f0, 0xe, 0x105504f38)
        /Users/maomao/go/pkg/mod/github.com/matrixorigin/matrixcube@v0.0.0-20210809144907-e1446fb0f0c7/server/server.go:76 +0x34
matrixone/pkg/vm/engine/aoe/dist.NewCubeDriverWithOptions(0x105501c98, 0x1400166ad80, 0x1054ef0f8, 0x14001dd21b0, 0x1054efb78, 0x140022c7350, 0x14001351940, 0xf, 0x14001351950, 0xf, ...)
        /Users/maomao/go/src/github.com/sukki37/matrixone/pkg/vm/engine/aoe/dist/cube_driver.go:208 +0x41c
matrixone/pkg/vm/engine/aoe/dist/testutil.NewTestClusterStore.func1(0x140005b54f0, 0x1400166ad80, 0x14001dd21b0, 0x140022c7350, 0x1400014e580, 0x14000256480)
        /Users/maomao/go/src/github.com/sukki37/matrixone/pkg/vm/engine/aoe/dist/testutil/testutil.go:97 +0xc4
created by matrixone/pkg/vm/engine/aoe/dist/testutil.NewTestClusterStore
        /Users/maomao/go/src/github.com/sukki37/matrixone/pkg/vm/engine/aoe/dist/testutil/testutil.go:95 +0x7f8

```

--- a/config/config.go
+++ b/config/config.go
@@ -88,6 +88,9 @@ type Config struct {
 	Customize CustomizeConfig
 	// Metric Config
 	Metric metric.Cfg `toml:"metric"`
+
+	// Test only used in testing
+	Test TestConfig
 }
 
 // Adjust adjust
@@ -134,6 +137,10 @@ func (c *Config) Adjust() {
 		c.Customize.CustomShardStateAwareFactory = func() aware.ShardStateAware {
 			return c.Customize.TestShardStateAware
 		}
+	}
+
+	if c.Test.Shards == nil {
+		c.Test.Shards = make(map[uint64]*TestShardConfig)
 	}
 }
 
@@ -397,4 +404,17 @@ type StoreHeartbeatDataProcessor interface {
 	HandleHeartbeatRsp(data []byte) error
 	// CollectData collect data at every heartbeat
 	CollectData() []byte
+}
+
+// TestConfig all test config
+type TestConfig struct {
+	Shards map[uint64]*TestShardConfig
+}
+
+// TestShardConfig shard test config
+type TestShardConfig struct {
+	// SkipSaveRaftApplyState skip save raft apply state to metastorage after applied a raft log
+	SkipSaveRaftApplyState bool
+	// SkipApply skip apply any raft log
+	SkipApply bool
 }

--- a/raftstore/peer_apply.go
+++ b/raftstore/peer_apply.go
@@ -201,6 +201,7 @@ type execResult struct {
 	changePeer   *changePeer
 	splitResult  *splitResult
 	raftGCResult *raftGCResult
+	needSyncData bool
 }
 
 type changePeer struct {
@@ -306,8 +307,12 @@ type applyDelegate struct {
 	pendingChangePeerCMD cmd
 	ctx                  *applyContext
 
-	// just for test
-	skipSaveRaftApplyState bool
+	// sync data after exec admin requests.
+	// Before restart we applied index is `100`, If `Customize.CustomAdjustInitAppliedIndexFactory` is set,
+	// after restart the init applied index maybe adjust to `10`. And raft will apply log again from [11, 100].
+	// If there has admin requests(e.g. `Split`, `ConfChange`) in [11, 100], we will encounter unknown errors.
+	// We use this flag to control whether Sync DataStorage is required after executing admin requests.
+	syncData bool
 }
 
 func (d *applyDelegate) clearAllCommandsAsStale() {
@@ -507,6 +512,10 @@ func (d *applyDelegate) applyConfChange(entry *raftpb.Entry) *execResult {
 }
 
 func (d *applyDelegate) doApplyRaftCMD() *execResult {
+	if sc, ok := d.store.cfg.Test.Shards[d.shard.ID]; ok && sc.SkipApply {
+		return nil
+	}
+
 	if d.ctx.index == 0 {
 		logger.Fatalf("shard %d apply raft command needs a none zero index",
 			d.shard.ID)
@@ -556,8 +565,10 @@ func (d *applyDelegate) doApplyRaftCMD() *execResult {
 	}
 
 	d.ctx.applyState.AppliedIndex = d.ctx.index
-	if !d.isPendingRemove() && !d.skipSaveRaftApplyState {
-		d.ctx.raftWB.Set(getRaftApplyStateKey(d.shard.ID), protoc.MustMarshal(&d.ctx.applyState))
+	if !d.isPendingRemove() {
+		if sc, ok := d.store.cfg.Test.Shards[d.shard.ID]; !ok || !sc.SkipSaveRaftApplyState {
+			d.ctx.raftWB.Set(getRaftApplyStateKey(d.shard.ID), protoc.MustMarshal(&d.ctx.applyState))
+		}
 	}
 
 	ds := d.store.DataStorageByGroup(d.shard.Group, d.shard.ID)
@@ -567,6 +578,13 @@ func (d *applyDelegate) doApplyRaftCMD() *execResult {
 			logger.Fatalf("shard %d commit apply result failed with %+v",
 				d.shard.ID,
 				err)
+		}
+	}
+
+	if d.syncData && result != nil && result.needSyncData {
+		err := d.store.DataStorageByGroup(d.shard.Group, d.shard.ID).Sync()
+		if err != nil {
+			logger.Fatalf("shard %d sync data failed with %+v")
 		}
 	}
 

--- a/raftstore/peer_apply_exec.go
+++ b/raftstore/peer_apply_exec.go
@@ -32,11 +32,23 @@ func (d *applyDelegate) execAdminRequest(ctx *applyContext) (*raftcmdpb.RaftCMDR
 	cmdType := ctx.req.AdminRequest.CmdType
 	switch cmdType {
 	case raftcmdpb.AdminCmdType_ChangePeer:
-		return d.doExecChangePeer(ctx)
+		resp, result, err := d.doExecChangePeer(ctx)
+		if result != nil {
+			result.needSyncData = true
+		}
+		return resp, result, err
 	case raftcmdpb.AdminCmdType_ChangePeerV2:
-		return d.doExecChangePeerV2(ctx)
+		resp, result, err := d.doExecChangePeerV2(ctx)
+		if result != nil {
+			result.needSyncData = true
+		}
+		return resp, result, err
 	case raftcmdpb.AdminCmdType_BatchSplit:
-		return d.doExecSplit(ctx)
+		resp, result, err := d.doExecSplit(ctx)
+		if result != nil {
+			result.needSyncData = true
+		}
+		return resp, result, err
 	case raftcmdpb.AdminCmdType_CompactLog:
 		return d.doExecCompactRaftLog(ctx)
 	}

--- a/raftstore/peer_apply_test.go
+++ b/raftstore/peer_apply_test.go
@@ -6,6 +6,8 @@ import (
 
 	"github.com/matrixorigin/matrixcube/components/prophet/util/typeutil"
 	"github.com/matrixorigin/matrixcube/config"
+	"github.com/matrixorigin/matrixcube/pb/bhmetapb"
+	"github.com/matrixorigin/matrixcube/storage/mem"
 	"github.com/stretchr/testify/assert"
 )
 
@@ -82,9 +84,11 @@ func TestIssue97(t *testing.T) {
 	assert.Equal(t, "OK", string(resps["w4"].Responses[0].Value))
 
 	// step4 write data to shard2 and only change shard2's raft local state
+	c.stores[0].cfg.Test.Shards[c.GetShardByID(5).ID] = &config.TestShardConfig{
+		SkipSaveRaftApplyState: true,
+	}
 	v, _ := c.stores[0].delegates.Load(c.GetShardByID(5).ID)
 	d := v.(*applyDelegate)
-	d.skipSaveRaftApplyState = true
 	d.ctx.raftWB.Reset()
 	resps, err = sendTestReqs(c.stores[0], time.Second*10, nil, nil,
 		createTestWriteReq("w5", "key2", "1"))
@@ -98,4 +102,56 @@ func TestIssue97(t *testing.T) {
 	_, err = sendTestReqs(c.stores[0], time.Second*10, nil, nil,
 		createTestReadReq("w5", "key2"))
 	assert.NoError(t, err)
+}
+
+func TestSyncData(t *testing.T) {
+	c := NewTestClusterStore(t,
+		DisableScheduleTestCluster,
+		SetCMDTestClusterHandler,
+		GetCMDTestClusterHandler,
+		WithAppendTestClusterAdjustConfigFunc(func(node int, cfg *config.Config) {
+			cfg.Replication.ShardCapacityBytes = typeutil.ByteSize(20)
+			cfg.Replication.ShardSplitCheckBytes = typeutil.ByteSize(10)
+			cfg.Customize.CustomAdjustInitAppliedIndexFactory = func(group uint64) func(shard bhmetapb.Shard, initAppliedIndex uint64) (adjustAppliedIndex uint64) {
+				return func(shard bhmetapb.Shard, initAppliedIndex uint64) (adjustAppliedIndex uint64) {
+					return initAppliedIndex
+				}
+			}
+		}))
+	defer c.Stop()
+
+	c.StartNode(0)
+	c.WaitLeadersByCount(t, 1, time.Second*10)
+	assert.Equal(t, uint64(0), c.dataStorages[0].(*mem.Storage).SyncCount)
+
+	cfg := c.stores[0].cfg
+	cfg.Replication.DisableShardSplit = true
+
+	// check change peer
+	c.StartNode(1)
+	c.WaitShardByCounts(t, [3]int{1, 1, 0}, time.Second*10)
+	assert.Equal(t, uint64(2), c.dataStorages[0].(*mem.Storage).SyncCount)
+
+	c.StartNode(2)
+	c.WaitShardByCounts(t, [3]int{1, 1, 1}, time.Second*10)
+	assert.Equal(t, uint64(4), c.dataStorages[0].(*mem.Storage).SyncCount)
+
+	// write key1
+	resps, err := sendTestReqs(c.stores[0], time.Second*10, nil, nil,
+		createTestWriteReq("w1", "key1", "value11"))
+	assert.NoError(t, err)
+	assert.Equal(t, "OK", string(resps["w1"].Responses[0].Value))
+	assert.Equal(t, uint64(4), c.dataStorages[0].(*mem.Storage).SyncCount)
+
+	// write key2
+	resps, err = sendTestReqs(c.stores[0], time.Second*10, nil, nil,
+		createTestWriteReq("w2", "key2", "value22"))
+	assert.NoError(t, err)
+	assert.Equal(t, "OK", string(resps["w2"].Responses[0].Value))
+	assert.Equal(t, uint64(4), c.dataStorages[0].(*mem.Storage).SyncCount)
+
+	// split
+	cfg.Replication.DisableShardSplit = false
+	c.WaitLeadersByCount(t, 2, time.Second*10)
+	assert.Equal(t, uint64(5), c.dataStorages[0].(*mem.Storage).SyncCount)
 }

--- a/raftstore/peer_job.go
+++ b/raftstore/peer_job.go
@@ -45,6 +45,8 @@ func (pr *peerReplica) startRegistrationJob() {
 		applyState:       pr.ps.raftApplyState,
 		appliedIndexTerm: pr.ps.appliedIndexTerm,
 		ctx:              newApplyContext(pr),
+		syncData: pr.store.cfg.Customize.CustomAdjustInitAppliedIndexFactory != nil &&
+			pr.store.cfg.Customize.CustomAdjustInitAppliedIndexFactory(pr.ps.shard.Group) != nil,
 	}
 
 	err := pr.store.addApplyJob(pr.applyWorker, "doRegistrationJob", func() error {

--- a/raftstore/peer_replica.go
+++ b/raftstore/peer_replica.go
@@ -462,7 +462,14 @@ func getRaftConfig(id, appliedIndex uint64, ps *peerStorage, cfg *config.Config)
 		if factory != nil {
 			newAppliedIndex := factory(ps.shard, appliedIndex)
 			if newAppliedIndex <= raftInitLogIndex {
-				newAppliedIndex = appliedIndex
+				if newAppliedIndex == 0 {
+					newAppliedIndex = appliedIndex
+				} else {
+					logger.Fatalf("shard %d unexpect adjust applied index, ajdust index %d must >= init applied index %d",
+						ps.shard.ID,
+						newAppliedIndex,
+						raftInitLogIndex)
+				}
 			}
 			if newAppliedIndex > appliedIndex {
 				logger.Fatalf("shard %d unexpect adjust applied index, ajdust index %d must <= real applied index %d",

--- a/raftstore/peer_replica.go
+++ b/raftstore/peer_replica.go
@@ -461,6 +461,9 @@ func getRaftConfig(id, appliedIndex uint64, ps *peerStorage, cfg *config.Config)
 		factory := cfg.Customize.CustomAdjustInitAppliedIndexFactory(ps.shard.Group)
 		if factory != nil {
 			newAppliedIndex := factory(ps.shard, appliedIndex)
+			if newAppliedIndex <= raftInitLogIndex {
+				newAppliedIndex = appliedIndex
+			}
 			if newAppliedIndex > appliedIndex {
 				logger.Fatalf("shard %d unexpect adjust applied index, ajdust index %d must <= real applied index %d",
 					ps.shard.ID,

--- a/raftstore/peer_storage.go
+++ b/raftstore/peer_storage.go
@@ -76,7 +76,7 @@ func newPeerStorage(store *store, shard bhmetapb.Shard) (*peerStorage, error) {
 	if err != nil {
 		return nil, err
 	}
-	logger.Infof("shard %d init raft apply state, state=<%+v>",
+	logger.Errorf("shard %d init raft apply state, state=<%+v>",
 		shard.ID,
 		s.raftApplyState)
 

--- a/raftstore/store.go
+++ b/raftstore/store.go
@@ -1043,7 +1043,7 @@ func (s *store) removeShardData(shard bhmetapb.Shard, job *task.Job) error {
 
 	start := encStartKey(&shard)
 	end := encEndKey(&shard)
-	return s.DataStorageByGroup(shard.Group, shard.ID).RemovedShardData(shard, start, end)
+	return s.DataStorageByGroup(shard.Group, shard.ID).RemoveShardData(shard, start, end)
 }
 
 // Delete all data that is not covered by `newShard`.

--- a/raftstore/testutil.go
+++ b/raftstore/testutil.go
@@ -20,6 +20,7 @@ import (
 	"testing"
 	"time"
 
+	cpebble "github.com/cockroachdb/pebble"
 	"github.com/fagongzi/log"
 	"github.com/matrixorigin/matrixcube/aware"
 	"github.com/matrixorigin/matrixcube/command"
@@ -73,14 +74,15 @@ var (
 type TestClusterOption func(*testClusterOptions)
 
 type testClusterOptions struct {
-	tmpDir            string
-	nodes             int
-	recreate          bool
-	adjustConfigFuncs []func(node int, cfg *config.Config)
-	storeFactory      func(node int, cfg *config.Config) Store
-	nodeStartFunc     func(node int, store Store)
-	logLevel          string
-	useDisk           bool
+	tmpDir             string
+	nodes              int
+	recreate           bool
+	adjustConfigFuncs  []func(node int, cfg *config.Config)
+	storeFactory       func(node int, cfg *config.Config) Store
+	nodeStartFunc      func(node int, store Store)
+	logLevel           string
+	useDisk            bool
+	dataOpts, metaOpts *cpebble.Options
 
 	writeHandlers map[uint64]command.WriteCommandFunc
 	readHandlers  map[uint64]command.ReadCommandFunc
@@ -91,6 +93,8 @@ type testClusterOptions struct {
 func newTestClusterOptions() *testClusterOptions {
 	return &testClusterOptions{
 		recreate:      true,
+		dataOpts:      &cpebble.Options{},
+		metaOpts:      &cpebble.Options{},
 		writeHandlers: make(map[uint64]command.WriteCommandFunc),
 		readHandlers:  make(map[uint64]command.ReadCommandFunc),
 	}
@@ -168,6 +172,20 @@ func WithTestClusterNodeStartFunc(value func(node int, store Store)) TestCluster
 func WithTestClusterRecreate(value bool) TestClusterOption {
 	return func(opts *testClusterOptions) {
 		opts.recreate = value
+	}
+}
+
+// WithDataStorageOption set options to create data storage
+func WithDataStorageOption(dataOpts *cpebble.Options) TestClusterOption {
+	return func(opts *testClusterOptions) {
+		opts.dataOpts = dataOpts
+	}
+}
+
+// WithMetadataStorageOption set options to create metadata storage
+func WithMetadataStorageOption(metaOpts *cpebble.Options) TestClusterOption {
+	return func(opts *testClusterOptions) {
+		opts.metaOpts = metaOpts
 	}
 }
 
@@ -466,7 +484,7 @@ func (c *TestRaftCluster) reset(opts ...TestClusterOption) {
 			var metaStorage storage.MetadataStorage
 			metaStorage = mem.NewStorage()
 			if c.opts.useDisk {
-				s, err := pebble.NewStorage(fmt.Sprintf("%s-meta", cfg.DataPath))
+				s, err := pebble.NewStorageWithOptions(fmt.Sprintf("%s-meta", cfg.DataPath), c.opts.metaOpts)
 				assert.NoError(c.t, err)
 				metaStorage = s
 			}
@@ -477,7 +495,7 @@ func (c *TestRaftCluster) reset(opts ...TestClusterOption) {
 			var dataStorage storage.DataStorage
 			dataStorage = mem.NewStorage()
 			if c.opts.useDisk {
-				s, err := pebble.NewStorage(fmt.Sprintf("%s-data", cfg.DataPath))
+				s, err := pebble.NewStorageWithOptions(fmt.Sprintf("%s-data", cfg.DataPath), c.opts.metaOpts)
 				assert.NoError(c.t, err)
 				dataStorage = s
 			}
@@ -512,7 +530,6 @@ func (c *TestRaftCluster) reset(opts ...TestClusterOption) {
 		c.stores = append(c.stores, s)
 		c.awares = append(c.awares, ts)
 	}
-
 }
 
 // EveryStore do every store, it can be used to init some store register
@@ -530,6 +547,16 @@ func (c *TestRaftCluster) GetStore(node int) Store {
 // GetWatcher returns the node store router watcher
 func (c *TestRaftCluster) GetWatcher(node int) prophet.Watcher {
 	return c.stores[node].router.(*defaultRouter).watcher
+}
+
+// StartNode start node
+func (c *TestRaftCluster) StartNode(node int) {
+	c.stores[node].Start()
+}
+
+// StopNode start node
+func (c *TestRaftCluster) StopNode(node int) {
+	c.stores[node].Stop()
 }
 
 // Start start the test raft cluster
@@ -557,10 +584,19 @@ func (c *TestRaftCluster) Start() {
 	wg.Wait()
 }
 
+// Restart restart the test cluster
 func (c *TestRaftCluster) Restart() {
+	c.RestartWithFunc(nil)
+}
+
+// RestartWithFunc restart the test cluster
+func (c *TestRaftCluster) RestartWithFunc(beforeStartFunc func()) {
 	c.Stop()
 	opts := append(c.initOpts, WithTestClusterRecreate(false))
 	c.reset(opts...)
+	if beforeStartFunc != nil {
+		beforeStartFunc()
+	}
 	c.Start()
 }
 
@@ -631,6 +667,16 @@ func (c *TestRaftCluster) WaitRemovedByShardID(t *testing.T, id uint64, timeout 
 	}
 }
 
+// GetShardLeaderStore returns the shard leader store
+func (c *TestRaftCluster) GetShardLeaderStore(id uint64) Store {
+	for idx, s := range c.stores {
+		if c.awares[idx].isLeader(id) {
+			return s
+		}
+	}
+	return nil
+}
+
 // WaitLeadersByCount check whether the number of shards leaders reaches a specific value until timeout
 func (c *TestRaftCluster) WaitLeadersByCount(t *testing.T, count int, timeout time.Duration) {
 	timeoutC := time.After(timeout)
@@ -649,16 +695,6 @@ func (c *TestRaftCluster) WaitLeadersByCount(t *testing.T, count int, timeout ti
 			time.Sleep(time.Millisecond * 100)
 		}
 	}
-}
-
-// GetShardLeaderStore returns the shard leader store
-func (c *TestRaftCluster) GetShardLeaderStore(id uint64) Store {
-	for idx, s := range c.stores {
-		if c.awares[idx].isLeader(id) {
-			return s
-		}
-	}
-	return nil
 }
 
 // WaitShardByCount check whether the number of shards reaches a specific value until timeout

--- a/storage/mem/storage.go
+++ b/storage/mem/storage.go
@@ -31,6 +31,9 @@ import (
 type Storage struct {
 	kv    *util.KVTree
 	stats stats.Stats
+
+	// SyncCount number of `Sync` method called
+	SyncCount uint64
 }
 
 // NewStorage returns a mem data storage
@@ -179,6 +182,12 @@ func (s *Storage) Seek(key []byte) ([]byte, []byte, error) {
 	return k, v, nil
 }
 
+// Sync sync data
+func (s *Storage) Sync() error {
+	atomic.AddUint64(&s.SyncCount, 1)
+	return nil
+}
+
 // Write write the data in batch
 func (s *Storage) Write(wb *util.WriteBatch, sync bool) error {
 	if len(wb.Ops) == 0 {
@@ -196,8 +205,8 @@ func (s *Storage) Write(wb *util.WriteBatch, sync bool) error {
 	return nil
 }
 
-// RemovedShardData remove shard data
-func (s *Storage) RemovedShardData(shard bhmetapb.Shard, encodedStartKey, encodedEndKey []byte) error {
+// RemoveShardData remove shard data
+func (s *Storage) RemoveShardData(shard bhmetapb.Shard, encodedStartKey, encodedEndKey []byte) error {
 	return s.RangeDelete(encodedStartKey, encodedEndKey)
 }
 

--- a/storage/pebble/storage.go
+++ b/storage/pebble/storage.go
@@ -32,6 +32,9 @@ import (
 type Storage struct {
 	db    *pebble.DB
 	stats stats.Stats
+
+	// SyncCount number of `Sync` method called
+	SyncCount uint64
 }
 
 // NewStorage returns pebble kv store on a default options
@@ -221,8 +224,8 @@ func (s *Storage) Free(pooled []byte) {
 
 }
 
-// RemovedShardData remove shard data
-func (s *Storage) RemovedShardData(shard bhmetapb.Shard, encodedStartKey, encodedEndKey []byte) error {
+// RemoveShardData remove shard data
+func (s *Storage) RemoveShardData(shard bhmetapb.Shard, encodedStartKey, encodedEndKey []byte) error {
 	return s.RangeDelete(encodedStartKey, encodedEndKey)
 }
 
@@ -333,6 +336,12 @@ func (s *Storage) Write(wb *util.WriteBatch, sync bool) error {
 	}
 
 	return s.db.Apply(b, opts)
+}
+
+// Sync sync data to disk
+func (s *Storage) Sync() error {
+	atomic.AddUint64(&s.SyncCount, 1)
+	return s.db.Flush()
 }
 
 // CreateSnapshot create a snapshot file under the giving path

--- a/storage/pebble/storage_test.go
+++ b/storage/pebble/storage_test.go
@@ -1,0 +1,47 @@
+package pebble
+
+import (
+	"os"
+	"testing"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	tmpDir = "/tmp/cube/storage/pebble"
+)
+
+func TestSync(t *testing.T) {
+	recreateTestTempDir(tmpDir)
+	opts := pebble.Options{}
+	opts.DisableWAL = true
+
+	s, err := NewStorageWithOptions("/tmp/cube/storage/pebble", &opts)
+	assert.NoError(t, err)
+
+	k := []byte("k")
+	v := []byte("v")
+	assert.NoError(t, s.Set(k, v))
+
+	s.Close()
+	s, err = NewStorageWithOptions("/tmp/cube/storage/pebble", &opts)
+	assert.NoError(t, err)
+	d, err := s.Get(k)
+	assert.NoError(t, err)
+	assert.Empty(t, d)
+	assert.NoError(t, s.Set(k, v))
+	assert.NoError(t, s.Sync())
+
+	s.Close()
+	s, err = NewStorageWithOptions("/tmp/cube/storage/pebble", &opts)
+	assert.NoError(t, err)
+	d, err = s.Get(k)
+	assert.NoError(t, err)
+	assert.Equal(t, v, d)
+}
+
+func recreateTestTempDir(tmpDir string) {
+	os.RemoveAll(tmpDir)
+	os.MkdirAll(tmpDir, 0755)
+}

--- a/storage/storage_data.go
+++ b/storage/storage_data.go
@@ -22,9 +22,10 @@ type DataStorage interface {
 	StatisticalStorage
 	CloseableStorage
 
-	// RemovedShardData remove shard data
-	RemovedShardData(shard bhmetapb.Shard, encodedStartKey, encodedEndKey []byte) error
-
+	// Sync sync data data to persistent device
+	Sync() error
+	// RemoveShardData remove shard data
+	RemoveShardData(shard bhmetapb.Shard, encodedStartKey, encodedEndKey []byte) error
 	// SplitCheck Find a key from [start, end), so that the sum of bytes of the value of [start, key) <=size,
 	// returns the current bytes in [start,end), and the founded key
 	SplitCheck(start []byte, end []byte, size uint64) (currentSize uint64, currentKeys uint64, splitKeys [][]byte, err error)

